[wip] Feat/refactoring cli lib

reubenmiller opened the pull request:
WIP: DO NOT MERGE!!!!

Testing new workflow action to run linting and tests. Documentation is still not finished yet

### Breaking changes

* Renamed command structure and parameter names. Checkout the [migration notes](docs/MIGRATION_V2.md) for full details to get the most out of the newer version.

* Restructured project and dependencies
    * Add code formatting and linting
    * Refactor code to simplify interactions between the local tcp server and the websocket client
    * Add unit and integration tests (coverage currently around ~85%)

* Enabled calling project using just the module name, i.e. `python3 -m c8ylp.main` => `python3 -m c8ylp`

* Added repo hosted cli [documentation](docs/cli/C8YLP.md)


### Features

* Load configuration from an environment via the `--env-file` parameter and support caching the Cumulocity token to it.

* Support usage of random port when starting the local proxy via `--port 0`

* Add interactive ssh connection which starts a once-off local proxy on a random port and then starts an SSH client automatically. Once the ssh client disconnects the local proxy is shutdown.

* [c8ylp plugin command](docs/cli/C8YLP_PLUGIN_COMMAND.md) Generic subcommand to execute custom commands to launch ssh yourself (instead of via `c8ylp connect ssh`). This allows you to customize your ssh options.

* Plugin support to allow the user to extend c8ylp functionality by use of external plugins in the form of python or shell scripts. See [docs/PLUGINS](docs/PLUGINS.md) for information

* Support multiple environment variables for setting Cumulocity credentials to improve compatibility with other Cumulocity tooling

### Bug fixes

* Port timeout error when running on windows

* Fixed randomized port race condition by letting the operating system assign a random port when starting the local tcp server

switschel commented:
@reubenmiller Found few things so far
1. dockerfile in devcontainer folder needs to be updated to be working again. Some missing files.
2. c8ylp login prompts always for TFA-Token even TFA for that user is not enabled. Cannot be skipped with empty values.

switschel commented:
Are there any reasons versioneer.py is copied into the project? Could we move it as a dependency in the requirements.txt?

reubenmiller commented:
> @reubenmiller Found few things so far
> 
> 1. dockerfile in devcontainer folder needs to be updated to be working again. Some missing files.
> 2. c8ylp login prompts always for TFA-Token even TFA for that user is not enabled. Cannot be skipped with empty values.

1. A `.devcontainer` folder was not added the project and I don't think was ever included. Or do you mean we should add one (more than happy to)?
2. Yeah I actually did wonder about that with the original implementation, but maybe I broke that. If so I can check out how I did it in go-c8y-cli and do something similar here.
3. Regarding versionier, that is unfortunately how it works. It needs to be committed to the project (as per the README.md, and https://github.com/python-versioneer/python-versioneer/issues/145)

reubenmiller commented:
@switschel optional TFA should be done now and I also added support for Basic Auth (only in cases where OAUTH2_INTERNAL has not been activated for some strange reason).

The only corner case which is not handled explicitly is if the tenant is using basic auth and TFA...but I'm not sure if it is worth while as we should encourage usage of OAUTH2_INTERNAL over basic.

Other than that regarding versioneer, I could change to using setuptools-scm instead...I have used it before, though some of the defaults are a bit strange imho.

switschel commented:
@reubenmiller 
1. devcontainer my fault.. missed it to add it to the repo.  I will add it with the required changes.
2. TFA Token: Will test this.
3. Versioneer: There might be license compatibility issues of CC0-1.0 with Apache v2 https://wiki.creativecommons.org/wiki/CC0_FAQ at least it is not clearly defined. We should check this before merging.

switschel commented:
@reubenmiller 
Can you add the copyright headers in all files? We have to follow the SAG OS guidelines where this is required.

reubenmiller commented:
@switschel
* Added a task which can add license (based on a template) to all .py files in the project. Added note to DEVELOPER.md docs.
* Removed versioneer in favor or setuptools_scm (which is installed as a dependency only). It might take one build to shake out any CI/CD bugs.

reubenmiller commented:
@switschel I've reworked the dev container, and fixed a bug (tasks have been added and updated in PR checklist). The only open topic now is the "Support TFA code for Basic Auth login type"....Though can cancel this task as it will be hard to test (as requires a SMS provider).

switschel commented:
@reubenmiller Looks fine now. 
One thing with the devcontainer: c8ylp is not installed correctly for me.
/bin/sh: 1: .devcontainer/entrypoint.sh: not found
[9087 ms] [14:25:52] Installing extension: njpwerner.autodocstring
[9147 ms] [14:25:52] Installing extension: ms-python.python
[9191 ms] [14:25:52] Installing extension: streetsidesoftware.code-spell-checker
[9204 ms] Error: Command failed: /bin/sh -c .devcontainer/entrypoint.sh
        at Xy (c:\Users\SWIT\.vscode\extensions\ms-vscode-remote.remote-containers-0.205.2\dist\extension\extension.js:300:15)
        at async Qy (c:\Users\SWIT\.vscode\extensions\ms-vscode-remote.remote-containers-0.205.2\dist\extension\extension.js:293:2645)
        at async c:\Users\SWIT\.vscode\extensions\ms-vscode-remote.remote-containers-0.205.2\dist\extension\extension.js:275:8257
